delta-lake/delta-spark330db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimizeExecutor.scala

```diff
@@ -0,0 +1,405 @@
+        }
+
+        bins.map(b => (partition, b))
+          // select bins that have at least two files or in case of multi-dim clustering
+          // select all bins
+          .filter(_._2.size > 1 || isMultiDimClustering)
+    }
+  }
+
+  /**
+   * Utility method to run a Spark job to compact the files in given bin
+   *
+   * @param txn [[OptimisticTransaction]] instance in use to commit the changes to DeltaLog.
+   * @param partition Partition values of the partition that files in [[bin]] belongs to.
+   * @param bin List of files to compact into one large file.
+   * @param maxFileSize Targeted output file size in bytes
+   */
+  private def runOptimizeBinJob(
+                                 txn: OptimisticTransaction,
+                                 partition: Map[String, String],
+                                 bin: Seq[AddFile],
+                                 maxFileSize: Long): Seq[FileAction] = {
+    val baseTablePath = txn.deltaLog.dataPath
+
+    val input = txn.deltaLog.createDataFrame(txn.snapshot, bin, actionTypeOpt = Some("Optimize"))
+    val repartitionDF = if (isMultiDimClustering) {
+      val totalSize = bin.map(_.size).sum
+      val approxNumFiles = Math.max(1, totalSize / maxFileSize).toInt
+      MultiDimClustering.cluster(
+        input,
+        approxNumFiles,
+        zOrderByColumns)
+    } else {
+      val useRepartition = sparkSession.sessionState.conf.getConf(
+        DeltaSQLConf.DELTA_OPTIMIZE_REPARTITION_ENABLED)
+      if (useRepartition) {
+        input.repartition(numPartitions = 1)
+      } else {
+        input.coalesce(numPartitions = 1)
+      }
+    }
+
+    val partitionDesc = partition.toSeq.map(entry => entry._1 + "=" + entry._2).mkString(",")
+
+    val partitionName = if (partition.isEmpty) "" else s" in partition ($partitionDesc)"
+    val description = s"$baseTablePath<br/>Optimizing ${bin.size} files" + partitionName
+    sparkSession.sparkContext.setJobGroup(
+      sparkSession.sparkContext.getLocalProperty(SPARK_JOB_GROUP_ID),
+      description)
+
+    val addFiles = txn.writeFiles(repartitionDF).collect {
+      case a: AddFile =>
+        a.copy(dataChange = false)
+      case other =>
+        throw new IllegalStateException(
+          s"Unexpected action $other with type ${other.getClass}. File compaction job output" +
+            s"should only have AddFiles")
+    }
+    val removeFiles = bin.map(f => f.removeWithTimestamp(operationTimestamp, dataChange = false))
+    val updates = addFiles ++ removeFiles
+    updates
+  }
+
+  type PartitionedBin = (Map[String, String], Seq[AddFile])
+
+  trait OptimizeType {
+    def minNumFiles: Long
+
+    def maxFileSize: Long =
+      sparkSession.sessionState.conf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_MAX_FILE_SIZE)
+
+    def targetFiles: (Seq[AddFile], Seq[AddFile])
+
+    def targetBins(jobs: Seq[PartitionedBin]): Seq[PartitionedBin] = jobs
+  }
+
+  case class Compaction() extends OptimizeType {
```

Comment:
If GpuOptimizeExecutor remains, do we really need these classes or can they be reused from the existing Delta Lake implementation on the platform?

Reply:
I had another go at this; it appears that these don't seem to have equivalents on Databricks that line up neatly. :/